### PR TITLE
Modify csproj and project.json files to use Contract references

### LINKED ISSUE #172: Make all test projects compile against contract rather than product source code
Currently all tests use ProjectReference to refer to the product code to test.  However, this permits use of API's not in the public contract.

CoreFx has adopted a model of annotating the ProjectReference with additional metadata to compile against the project.json contracts but still use the ProjectReference's build assembly to be used at runtime.  Look for "ReferenceOutputAssembly" in the CoreFx test .csproj's for examples.

This task involve adopting that convention for all WCF test projects.

### PR DESCRIPTION
**Pull request by iamjasonp**
Changes all our test projects to use contract references rather than P2P references. 

Not converted

````
src\System.Private.ServiceModel\tests\Scenarios\SelfHostWcfService\WcfService.csproj
````

Converted: 

````
src\System.Private.ServiceModel\tests\Common\src\System.ServiceModel.Tests.Common.csproj
src\System.Private.ServiceModel\tests\Common\tests\System.ServiceModel.Common.Tests.csproj
src\System.Private.ServiceModel\tests\Scenarios\Binding\Custom\Binding.Custom.Tests.csproj
src\System.Private.ServiceModel\tests\Scenarios\Binding\Http\Binding.Http.Tests.csproj
src\System.Private.ServiceModel\tests\Scenarios\Client\ChannelLayer\Client.ChannelLayer.Tests.csproj
src\System.Private.ServiceModel\tests\Scenarios\Client\ClientBase\Client.ClientBase.Tests.csproj
src\System.Private.ServiceModel\tests\Scenarios\Client\ExpectedExceptions\Client.ExpectedExceptions.Tests.csproj
src\System.Private.ServiceModel\tests\Scenarios\Client\TypedClient\Client.TypedClient.Tests.csproj
src\System.Private.ServiceModel\tests\Scenarios\Contract\Data\Contract.Data.Tests.csproj
src\System.Private.ServiceModel\tests\Scenarios\Contract\Fault\Contract.Fault.Tests.csproj
src\System.Private.ServiceModel\tests\Scenarios\Contract\Message\Contract.Message.Tests.csproj
src\System.Private.ServiceModel\tests\Scenarios\Contract\XmlSerializer\Contract.XmlSerializer.Tests.csproj
src\System.Private.ServiceModel\tests\Scenarios\Encoding\Encoders\Encoding.Encoders.Tests.csproj
src\System.Private.ServiceModel\tests\Scenarios\Encoding\MessageVersion\Encoding.MessageVersion.Tests.csproj
src\System.Private.ServiceModel\tests\Scenarios\Security\TransportSecurity\Security.TransportSecurity.Tests.csproj
src\System.Private.ServiceModel\tests\Scenarios\ServiceModel.Scenarios.Common\ServiceModel.Scenarios.Common.csproj
src\System.Private.ServiceModel\tests\Unit\System.Private.ServiceModel.Tests.csproj
src\System.ServiceModel.Duplex\tests\System.ServiceModel.Duplex.Tests.csproj
src\System.ServiceModel.Http\tests\System.ServiceModel.Http.Tests.csproj
src\System.ServiceModel.NetTcp\tests\System.ServiceModel.NetTcp.Tests.csproj
src\System.ServiceModel.Primitives\tests\System.ServiceModel.Primitives.Tests.csproj
src\System.ServiceModel.Security\tests\System.ServiceModel.Security.Tests.csproj
````

Fixes #172

### CODE REVIEW

**Comment by iamjasonp** [comment]
Note that this PR needs to be verified against the .NET Native and CoreCLR test environments prior to merge. This is on my todo list. 

I've built against OSS after a `git clean -fdx` and it seems to work.

File: src/System.Private.ServiceModel/tests/Common/src/project.json [hunk 1]
```diff
@@ -1,55 +1,10 @@
...
-    "System.IO": "4.0.10-beta-*",
-    "System.IO.Compression": "4.0.0-beta-*",
-    "System.Linq": "4.0.0-beta-*",
-    "System.Linq.Expressions": "4.0.10-beta-*",
-    "System.Linq.Queryable": "4.0.0-beta-*",
-    "System.Net.Http": "4.0.0-beta-*",
-    "System.Net.NameResolution": "4.0.0-beta-*",
-    "System.Net.Primitives": "4.0.10-beta-*",
-    "System.Net.Security": "4.0.0-beta-*",
-    "System.Net.Sockets": "4.0.10-beta-*",
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Common/src/project.json]: When I run OuterLoop tests, I see several failures due to not being able to find System.Net.Sockets or System.Reflection.Emit, etc.  I don't think we can remove the explicit dependencies on all the other packages. Did you find you needed to do this to build, or did you believe deleting these was a good cleanup?

File: src/System.Private.ServiceModel/tests/Common/src/project.lock.json [hunk 2]
```diff
@@ -367,15 +345,13 @@
           "lib/DNXCore50/System.Net.Security.dll": {}
         }
       },
-      "System.Net.Sockets/4.0.10-beta-23024": {
+      "System.Net.Sockets/4.0.0-beta-23024": {
```
Comment (roncain) [inline src/System.Private.ServiceModel/tests/Common/src/project.lock.json]: This downgrade to an earlier version of System.Net.Sockets is further evidence we want an explicit dependency in project.json.  I see that version '0' actually passes in NET Native but fails in CoreCLR.  I believe CoreCLR needs the '10' version.

File: src/System.Private.ServiceModel/tests/Common/src/project.json (hunk 1 above)
Comment (iamjasonp) [inline src/System.Private.ServiceModel/tests/Common/src/project.json]: Okay - this was something I anticipated but hadn't yet fixed (I was removing everything first and then adding back dependencies as needed).

**Comment by iamjasonp** [comment]
Tested the new changes in CoreCLR and .NET Native test infrastructure, with references to `System.Net.Sockets` and `System.Reflection.Emit.Lightweight` added back explicitly. Things seem to work. 

#196 opened as I think there's a further issue with `System.Private.ServiceModel` not referencing the correct nuget packages, but that's for another day. Once that happens, ac8df33 can be reverted.

**Comment by roncain** [comment]
LGTM.  I tried CoreCLR and NET Native test runs, and everything passed for me

**Comment by iamjasonp** [comment]
For an additional level of confidence, ran on another machine other than my main dev machine. Tests pass.

**Pull request merged**